### PR TITLE
feat: refactor preview context storage to support block index management

### PR DESCRIPTION
**Pull request by aichy126**
## Summary by CodeRabbit

* **Refactor**
  * Improved preview-context caching for more reliable and consistent preview behavior, with stricter validation and smarter pruning of stale entries.

* **Tests**
  * Added extensive tests covering preview-context persistence, truncation, backtracking, clearing rules, and message-pairing edge cases to ensure stable preview behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ai-shifu/ai-shifu/pull/1834?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/api/flaskr/service/learn/context_v2.py [hunk 1]
```diff
@@ -559,41 +563,130 @@ def clear(self) -> None:
...
         if not payload:
-            return []
-        if block_index == 0:
-            self.clear()
-            return []
+            return None
         doc_hash = payload.get("document_hash")
         if document and doc_hash and doc_hash != self._hash_document(document):
+            self.clear()
+            return None
```
[the remark this answers is not in the document]
Comment (aichy126) [reply src/api/flaskr/service/learn/context_v2.py:585]: Adopted. Tightened the hash check in `_load_entries` so any payload with `entries` but a missing/empty/mismatched `document_hash` is now treated as stale and cleared. Used the equivalent simpler form `if document and doc_hash != self._hash_document(document)` (drops the redundant `and doc_hash` short-circuit) — `!=` already covers `None` / empty / wrong-type cases. Added two regression tests (`test_missing_document_hash_clears_entries`, `test_empty_document_hash_clears_entries`). Fixed in f05b1f4f.

File: src/api/flaskr/service/learn/context_v2.py [hunk 2]
```diff
@@ -559,41 +563,130 @@ def clear(self) -> None:
...
+        entries = payload.get("entries")
```
[the remark this answers is not in the document]
Comment (aichy126) [reply src/api/flaskr/service/learn/context_v2.py:586]: Adopted (same root cause as the gemini-code-assist comment above). Applied the minimal equivalent — `if document and doc_hash != self._hash_document(document): self.clear(); return None` — instead of the suggested `isinstance(doc_hash, str)` branch, because Python's `!=` already invalidates `None` / empty string / wrong type against the computed sha256 hex. Two new tests cover missing and empty `document_hash` cases. Fixed in f05b1f4f.
